### PR TITLE
WINDUPRULE-982 Versions of Spring compatible with Jakarta EE 9

### PR DESCRIPTION
**Pull request by PhilipCattanach**
https://issues.redhat.com/browse/WINDUPRULE-982
A new ruleset that highlights versions of Spring and Spring Boot components that are not compatible with Jakarta EE 9

### CODE REVIEW

File: rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                    <dependency groupId="org.springframework.boot" artifactId="{*}" toVersion="2.9.9" />
+                    <project>
+                        <artifact groupId="org.springframework.boot" artifactId="{*}" toVersion="2.9.9"/>
+                    </project>
+                </or>
+            </when>
+            <perform>
+                <hint title="Version of Spring Boot not compatible with Jakarta EE 9+" effort="3" category-id="mandatory">
+                    <message>Version 3.0.0 is the minimum version of Spring Boot that is Jakarta EE 9+ compatible</message>
+                    <link href="https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6" title="A Java 17 and Jakarta EE 9 baseline for Spring Framework 6" />
```
Comment (mrizzi) [inline rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml:31]:
```suggestion
                    <link href="https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6/" title="A Java 17 and Jakarta EE 9 baseline for Spring Framework 6" />
```

File: rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                    <dependency groupId="org.springframework" artifactId="{*}" toVersion="5.9.9" />
+                    <project>
+                        <artifact groupId="org.springframework" artifactId="{*}" toVersion="5.9.9"/>
+                    </project>
+                </or>
+            </when>
+            <perform>
+                <hint title="Version of Spring not compatible with Jakarta EE 9+" effort="3" category-id="mandatory">
+                    <message>Version 6.0.0 is the minimum version of Spring that is Jakarta EE 9+ compatible</message>
+                    <link href="https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6" title="A Java 17 and Jakarta EE 9 baseline for Spring Framework 6" />
```
Comment (mrizzi) [inline rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml:48]:
```suggestion
                    <link href="https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6/" title="A Java 17 and Jakarta EE 9 baseline for Spring Framework 6" />
```

**Review by mrizzi — CHANGES_REQUESTED** [review]
The link test is failing because the `https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6`" URL redirectes with a `308` HTTP code to `https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6/` (`/` at the end).

@PhilipCattanach please accept the changes to have the links test to work.

File: rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        <description>
+            This ruleset ensures the versions of Spring and Spring Boot are Jakarta EE 9 compliant
+        </description>
+        <dependencies>
+            <addon id="org.jboss.windup.rules,windup-rules-javaee,3.0.0.Final" />
+            <addon id="org.jboss.windup.rules,windup-rules-java,3.0.0.Final" />
+            <addon id="org.jboss.windup.rules,windup-rules-xml,3.0.0.Final" />
+        </dependencies>
+        <targetTechnology id="jakarta-ee" versionRange="[9,)" />
+        <targetTechnology id="jws" versionRange="[6,)" />
```
Comment (mrizzi) [inline rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml:15]: I feel like we should consider having a `jws6` target alias (just like we have the `hibernate6` one) to define which are the relevant targets for such a migration (e.g. `jakarta-ee`).
In this way we could avoid "mixing" JWS migration with Spring/Spring Boot/Jakarta migrations

File: rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    </metadata>
+    <rules>
+        <rule id="spring-components-00001">
+            <!-- Spring Boot version 3.0.0 is Jakarta compliant -->
+            <when>
+                <or>
+                    <dependency groupId="org.springframework.boot" artifactId="{*}" toVersion="2.9.9" />
```
Comment (mrizzi) [inline rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml:22]: I can see the latest SB version is `2.7.7` but we could be more defensive in the `toVersion` maybe having something like `2.99.999` to be deliberately high with the version to ensure we will be able to cover many further SB 2 maintenance release

File: rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                    <message>Version 3.0.0 is the minimum version of Spring Boot that is Jakarta EE 9+ compatible</message>
+                    <link href="https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6/" title="A Java 17 and Jakarta EE 9 baseline for Spring Framework 6" />
+                </hint>
+            </perform>
+        </rule>
+                <rule id="spring-components-00001">
+            <!-- Spring version 6.0.0 is Jakarta compliant -->
+            <when>
+                <or>
+                    <dependency groupId="org.springframework" artifactId="{*}" toVersion="5.9.9" />
```
Comment (mrizzi) [inline rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml:39]: Same as above, maybe having `5.99.999` is safer.

**Comment by PhilipCattanach** [comment]
@mrizzi following the JWS6 meeting I have updated the jws6 technology transformer to include hibernate 6
I tested by building the CLI and running this analysis to ensure the hibernate 6 rules fired.

./windup-cli --input /home/pcattana/Repositories/windup-rulesets/rules/rules-reviewed/eap8/eap7/tests/data/hibernate6 --target jws6 --output $HOME/MTA_Output/JWS6Transformer --sourceMode

File: rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml (hunk 4 above)
Comment (PhilipCattanach) [reply rules/rules-reviewed/jakarta-ee9/java-ee/spring-components.windup.xml]: more defensive versioning applied

**Review by mrizzi — APPROVED** [review]
LGTM, thanks @PhilipCattanach :+1:

**Pull request merged**